commit 2d98b4fb0ae49aa22c735eda6f221da91b4bc3cb
Author:     Nikolay Sivov <nsivov@codeweavers.com>
AuthorDate: Tue Aug 11 19:36:00 2026 +0200
Commit:     Alexandre Julliard <julliard@winehq.org>
CommitDate: Wed Aug 19 17:23:59 2026 +0200

    d3d10/tests: Add some state block capture tests with disabled fields.
    
    Signed-off-by: Nikolay Sivov <nsivov@codeweavers.com>

commit a5c665c30fe3d1ba6748680e68e77a564d58a9d4
Author:     Nikolay Sivov <nsivov@codeweavers.com>
AuthorDate: Tue Aug 11 19:53:10 2026 +0200
Commit:     Alexandre Julliard <julliard@winehq.org>
CommitDate: Wed Aug 19 17:23:59 2026 +0200

    d3d10/stateblock: Fix mask checks for byte-sized fields.
    
    Disabling such states toggles least significant bit only,
    remaining bits stay intact.
    
    Signed-off-by: Nikolay Sivov <nsivov@codeweavers.com>

Runeon backport adaptation: Combine the upstream test prerequisite and fix into one net diff; retain authorship and all upstream test assertions. Add an explicit no-device skip before using the D3D10 device. No production-source semantic adaptation.

diff --git a/dlls/d3d10/stateblock.c b/dlls/d3d10/stateblock.c
--- a/dlls/d3d10/stateblock.c
+++ b/dlls/d3d10/stateblock.c
@@ -316,6 +316,14 @@
     return refcount;
 }
 
+static BOOL stateblock_mask_get_bit(BYTE *field, UINT field_size, UINT idx)
+{
+    if (idx >= field_size)
+        return FALSE;
+
+    return field[idx >> 3] & (1 << (idx & 7));
+}
+
 static HRESULT STDMETHODCALLTYPE d3d10_stateblock_Capture(ID3D10StateBlock *iface)
 {
     unsigned int vp_count = D3D10_VIEWPORT_AND_SCISSORRECT_OBJECT_COUNT_PER_PIPELINE;
@@ -326,7 +334,7 @@
 
     stateblock_cleanup(stateblock);
 
-    if (stateblock->mask.VS)
+    if (stateblock_mask_get_bit(&stateblock->mask.VS, 1, 0))
         ID3D10Device_VSGetShader(stateblock->device, &stateblock->vs);
     for (i = 0; i < D3D10_COMMONSHADER_SAMPLER_SLOT_COUNT; ++i)
     {
@@ -344,7 +352,7 @@
             ID3D10Device_VSGetConstantBuffers(stateblock->device, i, 1, &stateblock->vs_cbs[i]);
     }
 
-    if (stateblock->mask.GS)
+    if (stateblock_mask_get_bit(&stateblock->mask.GS, 1, 0))
         ID3D10Device_GSGetShader(stateblock->device, &stateblock->gs);
     for (i = 0; i < D3D10_COMMONSHADER_SAMPLER_SLOT_COUNT; ++i)
     {
@@ -362,7 +370,7 @@
             ID3D10Device_GSGetConstantBuffers(stateblock->device, i, 1, &stateblock->gs_cbs[i]);
     }
 
-    if (stateblock->mask.PS)
+    if (stateblock_mask_get_bit(&stateblock->mask.PS, 1, 0))
         ID3D10Device_PSGetShader(stateblock->device, &stateblock->ps);
     for (i = 0; i < D3D10_COMMONSHADER_SAMPLER_SLOT_COUNT; ++i)
     {
@@ -386,35 +394,35 @@
             ID3D10Device_IAGetVertexBuffers(stateblock->device, i, 1, &stateblock->vbs[i],
                     &stateblock->vb_strides[i], &stateblock->vb_offsets[i]);
     }
-    if (stateblock->mask.IAIndexBuffer)
+    if (stateblock_mask_get_bit(&stateblock->mask.IAIndexBuffer, 1, 0))
         ID3D10Device_IAGetIndexBuffer(stateblock->device, &stateblock->ib,
                 &stateblock->ib_format, &stateblock->ib_offset);
-    if (stateblock->mask.IAInputLayout)
+    if (stateblock_mask_get_bit(&stateblock->mask.IAInputLayout, 1, 0))
         ID3D10Device_IAGetInputLayout(stateblock->device, &stateblock->il);
-    if (stateblock->mask.IAPrimitiveTopology)
+    if (stateblock_mask_get_bit(&stateblock->mask.IAPrimitiveTopology, 1, 0))
         ID3D10Device_IAGetPrimitiveTopology(stateblock->device, &stateblock->topology);
 
-    if (stateblock->mask.OMRenderTargets)
+    if (stateblock_mask_get_bit(&stateblock->mask.OMRenderTargets, 1, 0))
         ID3D10Device_OMGetRenderTargets(stateblock->device, D3D10_SIMULTANEOUS_RENDER_TARGET_COUNT,
                 stateblock->rtvs, &stateblock->dsv);
-    if (stateblock->mask.OMDepthStencilState)
+    if (stateblock_mask_get_bit(&stateblock->mask.OMDepthStencilState, 1, 0))
         ID3D10Device_OMGetDepthStencilState(stateblock->device, &stateblock->dss, &stateblock->stencil_ref);
-    if (stateblock->mask.OMBlendState)
+    if (stateblock_mask_get_bit(&stateblock->mask.OMBlendState, 1, 0))
         ID3D10Device_OMGetBlendState(stateblock->device, &stateblock->bs,
                 stateblock->blend_factor, &stateblock->sample_mask);
 
-    if (stateblock->mask.RSViewports)
+    if (stateblock_mask_get_bit(&stateblock->mask.RSViewports, 1, 0))
         ID3D10Device_RSGetViewports(stateblock->device, &vp_count, stateblock->vps);
-    if (stateblock->mask.RSScissorRects)
+    if (stateblock_mask_get_bit(&stateblock->mask.RSScissorRects, 1, 0))
         ID3D10Device_RSGetScissorRects(stateblock->device, &vp_count, stateblock->scissor_rects);
-    if (stateblock->mask.RSRasterizerState)
+    if (stateblock_mask_get_bit(&stateblock->mask.RSRasterizerState, 1, 0))
         ID3D10Device_RSGetState(stateblock->device, &stateblock->rs);
 
-    if (stateblock->mask.SOBuffers)
+    if (stateblock_mask_get_bit(&stateblock->mask.SOBuffers, 1, 0))
         ID3D10Device_SOGetTargets(stateblock->device, D3D10_SO_BUFFER_SLOT_COUNT,
                 stateblock->so_buffers, stateblock->so_offsets);
 
-    if (stateblock->mask.Predication)
+    if (stateblock_mask_get_bit(&stateblock->mask.Predication, 1, 0))
         ID3D10Device_GetPredication(stateblock->device, &stateblock->predicate, &stateblock->predicate_value);
 
     return S_OK;
@@ -427,7 +435,7 @@
 
     TRACE("iface %p.\n", iface);
 
-    if (stateblock->mask.VS)
+    if (stateblock_mask_get_bit(&stateblock->mask.VS, 1, 0))
         ID3D10Device_VSSetShader(stateblock->device, stateblock->vs);
     for (i = 0; i < D3D10_COMMONSHADER_SAMPLER_SLOT_COUNT; ++i)
     {
@@ -445,7 +453,7 @@
             ID3D10Device_VSSetConstantBuffers(stateblock->device, i, 1, &stateblock->vs_cbs[i]);
     }
 
-    if (stateblock->mask.GS)
+    if (stateblock_mask_get_bit(&stateblock->mask.GS, 1, 0))
         ID3D10Device_GSSetShader(stateblock->device, stateblock->gs);
     for (i = 0; i < D3D10_COMMONSHADER_SAMPLER_SLOT_COUNT; ++i)
     {
@@ -463,7 +471,7 @@
             ID3D10Device_GSSetConstantBuffers(stateblock->device, i, 1, &stateblock->gs_cbs[i]);
     }
 
-    if (stateblock->mask.PS)
+    if (stateblock_mask_get_bit(&stateblock->mask.PS, 1, 0))
         ID3D10Device_PSSetShader(stateblock->device, stateblock->ps);
     for (i = 0; i < D3D10_COMMONSHADER_SAMPLER_SLOT_COUNT; ++i)
     {
@@ -487,37 +495,37 @@
             ID3D10Device_IASetVertexBuffers(stateblock->device, i, 1, &stateblock->vbs[i],
                     &stateblock->vb_strides[i], &stateblock->vb_offsets[i]);
     }
-    if (stateblock->mask.IAIndexBuffer)
+    if (stateblock_mask_get_bit(&stateblock->mask.IAIndexBuffer, 1, 0))
         ID3D10Device_IASetIndexBuffer(stateblock->device, stateblock->ib,
                 stateblock->ib_format, stateblock->ib_offset);
-    if (stateblock->mask.IAInputLayout)
+    if (stateblock_mask_get_bit(&stateblock->mask.IAInputLayout, 1, 0))
         ID3D10Device_IASetInputLayout(stateblock->device, stateblock->il);
-    if (stateblock->mask.IAPrimitiveTopology)
+    if (stateblock_mask_get_bit(&stateblock->mask.IAPrimitiveTopology, 1, 0))
         ID3D10Device_IASetPrimitiveTopology(stateblock->device, stateblock->topology);
 
-    if (stateblock->mask.OMRenderTargets)
+    if (stateblock_mask_get_bit(&stateblock->mask.OMRenderTargets, 1, 0))
         ID3D10Device_OMSetRenderTargets(stateblock->device, D3D10_SIMULTANEOUS_RENDER_TARGET_COUNT,
                 stateblock->rtvs, stateblock->dsv);
-    if (stateblock->mask.OMDepthStencilState)
+    if (stateblock_mask_get_bit(&stateblock->mask.OMDepthStencilState, 1, 0))
         ID3D10Device_OMSetDepthStencilState(stateblock->device, stateblock->dss, stateblock->stencil_ref);
-    if (stateblock->mask.OMBlendState)
+    if (stateblock_mask_get_bit(&stateblock->mask.OMBlendState, 1, 0))
         ID3D10Device_OMSetBlendState(stateblock->device, stateblock->bs,
                 stateblock->blend_factor, stateblock->sample_mask);
 
-    if (stateblock->mask.RSViewports)
+    if (stateblock_mask_get_bit(&stateblock->mask.RSViewports, 1, 0))
         ID3D10Device_RSSetViewports(stateblock->device, D3D10_VIEWPORT_AND_SCISSORRECT_OBJECT_COUNT_PER_PIPELINE,
                 stateblock->vps);
-    if (stateblock->mask.RSScissorRects)
+    if (stateblock_mask_get_bit(&stateblock->mask.RSScissorRects, 1, 0))
         ID3D10Device_RSSetScissorRects(stateblock->device, D3D10_VIEWPORT_AND_SCISSORRECT_OBJECT_COUNT_PER_PIPELINE,
                 stateblock->scissor_rects);
-    if (stateblock->mask.RSRasterizerState)
+    if (stateblock_mask_get_bit(&stateblock->mask.RSRasterizerState, 1, 0))
         ID3D10Device_RSSetState(stateblock->device, stateblock->rs);
 
-    if (stateblock->mask.SOBuffers)
+    if (stateblock_mask_get_bit(&stateblock->mask.SOBuffers, 1, 0))
         ID3D10Device_SOSetTargets(stateblock->device, D3D10_SO_BUFFER_SLOT_COUNT,
                 stateblock->so_buffers, stateblock->so_offsets);
 
-    if (stateblock->mask.Predication)
+    if (stateblock_mask_get_bit(&stateblock->mask.Predication, 1, 0))
         ID3D10Device_SetPredication(stateblock->device, stateblock->predicate, stateblock->predicate_value);
 
     return S_OK;
@@ -574,14 +582,6 @@
     *stateblock = &object->ID3D10StateBlock_iface;
 
     return S_OK;
-}
-
-static BOOL stateblock_mask_get_bit(BYTE *field, UINT field_size, UINT idx)
-{
-    if (idx >= field_size)
-        return FALSE;
-
-    return field[idx >> 3] & (1 << (idx & 7));
 }
 
 static HRESULT stateblock_mask_set_bits(BYTE *field, UINT field_size, UINT start_bit, UINT count)
diff --git a/dlls/d3d10/tests/device.c b/dlls/d3d10/tests/device.c
--- a/dlls/d3d10/tests/device.c
+++ b/dlls/d3d10/tests/device.c
@@ -19,6 +19,20 @@
 #define COBJMACROS
 #include "d3d10.h"
 #include "wine/test.h"
+
+static ID3D10Device *create_device(void)
+{
+    ID3D10Device *device;
+
+    if (SUCCEEDED(D3D10CreateDevice(NULL, D3D10_DRIVER_TYPE_HARDWARE, NULL, 0, D3D10_SDK_VERSION, &device)))
+        return device;
+    if (SUCCEEDED(D3D10CreateDevice(NULL, D3D10_DRIVER_TYPE_WARP, NULL, 0, D3D10_SDK_VERSION, &device)))
+        return device;
+    if (SUCCEEDED(D3D10CreateDevice(NULL, D3D10_DRIVER_TYPE_REFERENCE, NULL, 0, D3D10_SDK_VERSION, &device)))
+        return device;
+
+    return NULL;
+}
 
 static void test_create_device(void)
 {
@@ -200,8 +214,100 @@
     ok(!ret, "Got unexpected ret %#x.\n", ret);
 }
 
+static void test_stateblock(void)
+{
+#if 0
+float4 main(float4 pos : POSITION) : POSITION
+{
+    return pos;
+}
+#endif
+    static const DWORD simple_vs[] =
+    {
+        0x43425844, 0x66689e7c, 0x643f0971, 0xb7f67ff4, 0xabc48688, 0x00000001, 0x000000d4, 0x00000003,
+        0x0000002c, 0x00000060, 0x00000094, 0x4e475349, 0x0000002c, 0x00000001, 0x00000008, 0x00000020,
+        0x00000000, 0x00000000, 0x00000003, 0x00000000, 0x00000f0f, 0x49534f50, 0x4e4f4954, 0xababab00,
+        0x4e47534f, 0x0000002c, 0x00000001, 0x00000008, 0x00000020, 0x00000000, 0x00000000, 0x00000003,
+        0x00000000, 0x0000000f, 0x49534f50, 0x4e4f4954, 0xababab00, 0x52444853, 0x00000038, 0x00010040,
+        0x0000000e, 0x0300005f, 0x001010f2, 0x00000000, 0x03000065, 0x001020f2, 0x00000000, 0x05000036,
+        0x001020f2, 0x00000000, 0x00101e46, 0x00000000, 0x0100003e,
+    };
+
+    ID3D10Buffer *index_buffer, *buffer;
+    ID3D10VertexShader *vs, *tmp_vs;
+    D3D10_BUFFER_DESC buffer_desc;
+    ID3D10StateBlock *state_block;
+    D3D10_STATE_BLOCK_MASK mask;
+    float blend_factor[4];
+    ID3D10Device *device;
+    DXGI_FORMAT format;
+    UINT offset;
+    HRESULT hr;
+
+    if (!(device = create_device()))
+    {
+        skip("Failed to create a D3D10 device.\n");
+        return;
+    }
+
+    hr = ID3D10Device_CreateVertexShader(device, simple_vs, sizeof(simple_vs), &vs);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    buffer_desc.ByteWidth = 64;
+    buffer_desc.Usage = D3D10_USAGE_DEFAULT;
+    buffer_desc.BindFlags = D3D10_BIND_INDEX_BUFFER;
+    buffer_desc.CPUAccessFlags = 0;
+    buffer_desc.MiscFlags = 0;
+    hr = ID3D10Device_CreateBuffer(device, &buffer_desc, NULL, &index_buffer);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = D3D10StateBlockMaskEnableAll(&mask);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = D3D10StateBlockMaskDisableCapture(&mask, D3D10_DST_OM_BLEND_STATE, 0, 1);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = D3D10StateBlockMaskDisableCapture(&mask, D3D10_DST_VS, 0, 1);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = D3D10StateBlockMaskDisableCapture(&mask, D3D10_DST_IA_INDEX_BUFFER, 0, 1);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    memset(&blend_factor, 0, sizeof(blend_factor));
+    blend_factor[0] = 0.1f;
+    ID3D10Device_OMSetBlendState(device, NULL, blend_factor, 0);
+    ID3D10Device_VSSetShader(device, vs);
+    ID3D10Device_IASetIndexBuffer(device, index_buffer, DXGI_FORMAT_R16_UINT, 0);
+
+    hr = D3D10CreateStateBlock(device, &mask, &state_block);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = state_block->lpVtbl->Capture(state_block);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    memset(&blend_factor, 0, sizeof(blend_factor));
+    blend_factor[0] = 0.2f;
+    ID3D10Device_OMSetBlendState(device, NULL, blend_factor, 0);
+    ID3D10Device_VSSetShader(device, NULL);
+    ID3D10Device_IASetIndexBuffer(device, NULL, 0, 0);
+
+    hr = state_block->lpVtbl->Apply(state_block);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    memset(&blend_factor, 0, sizeof(blend_factor));
+    ID3D10Device_OMGetBlendState(device, NULL, blend_factor, NULL);
+    ok(blend_factor[0] == 0.2f, "Unexpected factor %.8e.\n", blend_factor[0]);
+    ID3D10Device_VSGetShader(device, &tmp_vs);
+    ok(!tmp_vs, "Unexpected object %p.\n", tmp_vs);
+    ID3D10Device_IAGetIndexBuffer(device, &buffer, &format, &offset);
+    ok(!buffer, "Unexpected object %p.\n", buffer);
+
+    ID3D10Buffer_Release(index_buffer);
+    ID3D10StateBlock_Release(state_block);
+    ID3D10VertexShader_Release(vs);
+    ID3D10Device_Release(device);
+}
+
 START_TEST(device)
 {
     test_create_device();
     test_stateblock_mask();
-}
+    test_stateblock();
+}
